test/test-parser.js

```diff
@@ -85,115 +87,214 @@ const assertThrows = (callback, expectedMessage, options) => {
   assert(thrown, 'no error was thrown');
 };
 
-const wrapper = (instance, strings, language) => {
-  const bindings = {
-    boolean: [],
-    defined: [],
-    attribute: [],
-    property: [],
-    content: [],
-    text: [],
+// Simpler helper to add some test information under a TEST symbol.
+const wrapper = strings => {
+  const tokens = [];
+  const onToken = (type, index, start, end, substring) => {
+    tokens.push({ type, index, start, end, substring });
   };
-  const onBoolean = (name, path) => bindings.boolean.push({ name, path: String(path) });
-  const onDefined = (name, path) => bindings.defined.push({ name, path: String(path) });
-  const onAttribute = (name, path) => bindings.attribute.push({ name, path: String(path) });
-  const onProperty = (name, path) => bindings.property.push({ name, path: String(path) });
-  const onContent = path => bindings.content.push({ path: String(path) });
-  const onText = path => bindings.text.push({ path: String(path) });
-  const fragment = instance.parse(strings, onBoolean, onDefined, onAttribute, onProperty, onContent, onText, language);
-  fragment[TEST] = { bindings };
-  return fragment;
+  XParser.parse(strings, onToken);
+  tokens[TEST] = { strings };
+  return tokens;
 };
 
 // Simple function to return strings array from tagged template function call.
 //  Since IDEs will pick up on the “html” syntax, they should highlight, which
 //  will make this more readable.
-const html = strings => wrapper(parser, strings);
+const html = strings => wrapper(strings);
 
 // Certain tests require that we write _terribly broken_ html. To prevent IDEs
 //  from choking when trying to highlight, we also have a “htmlol” function.
-const htmlol = strings => wrapper(parser, strings);
-
-// Placeholder for values which delimit tagged template functions. These are
-//  not considered by the parser, so the value here has no significance and will
-//  not show up in any of the resulting fragments from the initial parse.
-const VALUE = '…';
+const htmlol = strings => wrapper(strings);
 
 describe('basic', () => {
   it('empty template works', () => {
-    const fragment = html``;
-    assert(fragment.childNodes.length === 0);
+    const expectedTokens = [];
+    const tokens = html``;
+    assert(deepEqual(tokens, expectedTokens), stringifyTokens(tokens));
   });
 
   it('single string works', () => {
-    const fragment = html`<div id="target">No interpolation.</div>`;
-    assert(fragment.childNodes.length === 1);
-    assert(fragment.querySelector('#target').textContent === 'No interpolation.');
+    const expectedTokens = [
```

Comment:
The test suite for the parser is more verbose now, but it’s also very exhaustive. It’s quite clear now that we’re interpreting the interpreted markup as expected.